### DISCUSSION TITLE
Hyperscan Preset changes to build Chimera

### DISCUSSION THREAD

**Discussion by simpleyetawesomeprojects** [bytedeco/javacpp-presets#1194] [Q&A]
Hello Community,

Thank you for creating the hyperscan preset, which is very helpful. Hperscan is excellent, but it does not have full support for PCRE syntax, and Chimera which is included in hyperscan source has full support for PCRE syntax. Chimera is a hybrid of Hyperscan and PCRE. 

Per the documentation, http://intel.github.io/hyperscan/dev-reference/chimera.html, 

_"Building Hyperscan will automatically generate Chimera library. Currently only static library is supported for Chimera, so please use static build type when configure CMake build options"_

The hyperscan preset does not include changes for Chimera and hence it does not build Chimera. Could you please let me know if there is any plan to incorporate changes to build Chimera? 

I have been experimenting with building the changes in the below PR, I'm a newbie to javacpp and unable to figure out what needs to be done to make Chimera work.

The PR includes a Chimera Test example class, when I run it, I see the below error. How can I resolve this error? Any help is greatly appreciated.

`dyld: lazy symbol binding failed: Symbol not found: _ch_compile_multi
  Referenced from: /Users/aaa/.javacpp/cache/hyperscan-5.4.0-1.5.8-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/hyperscan/macosx-x86_64/libjnihyperscan.dylib
  Expected in: flat namespace

dyld: Symbol not found: _ch_compile_multi
  Referenced from: /Users/aaa/.javacpp/cache/hyperscan-5.4.0-1.5.8-SNAPSHOT-macosx-x86_64.jar/org/bytedeco/hyperscan/macosx-x86_64/libjnihyperscan.dylib
  Expected in: flat namespace`

here is the PR with the changes to build Chimera
https://github.com/simpleyetawesomeprojects/javacpp-presets/pull/1

**Comment by saudet** [comment]
It sounds like we need to link with an additional library for Chimera. Try to do a `find -name *.a` and place all the names you find there in the `@Platform(link=...` list here: https://github.com/bytedeco/javacpp-presets/blob/master/hyperscan/src/main/java/org/bytedeco/hyperscan/presets/hyperscan.java#L40

**Reply by simpleyetawesomeprojects** [reply]
I added all the static linker file names, after the changes the existing "HyperscanTest" example code does not work anymore. This class and the new "ChScanExample" both have runtime errors. It looks like the lib files were not added automatically into the platform jar after linking the ".a" files.

<img width="1253" alt="Screen Shot 2022-06-27 at 3 41 52 PM" src="https://user-images.githubusercontent.com/107608079/176048189-5392374c-1df6-4c47-8bed-e561daa564cc.png">



And, these are all the build files
<img width="349" alt="Screen Shot 2022-06-27 at 3 34 51 PM" src="https://user-images.githubusercontent.com/107608079/176048019-aad21980-bde0-4275-838d-1e751ac53b6a.png">



here is the updated pr
https://github.com/simpleyetawesomeprojects/javacpp-presets/pull/1/files#diff-0a8717e5623b2d50da7c1a25e1f5246a7019aed8ee0e73cc497383b45467a263R45

**Reply by saudet** [reply]
Static libraries don't need to be bundled in JAR files since they get linked with the JNI one, so that part is fine. In any case, it's something related to linking. We'll need to figure out which libraries exactly we should be linking to. Make sure to check in the documentation to see if they mention anything about that, and if they do, then we can simply add the recommended ones to the list for JavaCPP. If they don't, someone is going to need to figure this out by trial and error...

**Reply by simpleyetawesomeprojects** [reply]
The documentation says

`**Requirements
The PCRE library (http://pcre.org/) version 8.41 is required for Chimera.

Note
Since Chimera needs to reference PCRE internal function, please place PCRE source directory under Hyperscan root directory in order to build Chimera.

Beside this, both hardware and software requirements of Chimera are the same to Hyperscan. See [Hardware](http://intel.github.io/hyperscan/dev-reference/getting_started.html#hardware) and [Software](http://intel.github.io/hyperscan/dev-reference/getting_started.html#software) for more details.

Note
Building Hyperscan will automatically generate Chimera library. Currently only static library is supported for Chimera, so please use static build type when configure CMake build options.**`

I've incorporated these steps. As you mentioned, it's the linking part that I'm unable to resolve. I tried a few trials with no luck and finally reached out to see if I can get some help. 

Thanks for your responses though. I appreciate your time.

**Reply by saudet** [reply]
Since it says it only works with static libraries, make sure to disable the creation of **all** shared libraries.
